contracts/protocol/integration/amm/ArrakisUniswapV3AmmAdapter.sol

```diff
@@ -0,0 +1,277 @@
+        address[] calldata _components,
+        uint256[] calldata _minTokensOut,
+        uint256 _liquidity
+    )
+        external
+        view
+        override
+        returns (address target, uint256 value, bytes memory data)
+    {   
+        address setToken = _setToken;
+        address[] memory components = _components;
+        uint256[] memory minTokensOut = _minTokensOut;
+        uint256 liquidity = _liquidity;
+        IArrakisVaultV1 arrakisVaultPool = IArrakisVaultV1(_pool);
+
+        // Make sure that only up to the amount of liquidity tokens owned by the Set Token are redeemed
+        uint256 setTokenLiquidityBalance = arrakisVaultPool.balanceOf(setToken);
+        require(liquidity <= setTokenLiquidityBalance, "_liquidity must be <= to current balance");
+
+        // Checks for minTokensOut
+        require(minTokensOut[0] > 0 && minTokensOut[1] > 0, "Minimum quantity must be nonzero");
+
+        // Sort the amount in order of tokens stored in Arrakis Pool
+        (uint256 minTokensOutA, uint256 minTokensOutB) = _getOrderedAmount(components[0], components[1], minTokensOut[0], minTokensOut[1]);
+
+        target = router;
+        value = 0;
+        data = abi.encodeWithSignature(
+            REMOVE_LIQUIDITY,
+            arrakisVaultPool,
+            liquidity,
+            minTokensOutA,
+            minTokensOutB,
+            setToken
+        );
+    }
+
+    /**
+     * Return calldata for the remove liquidity single asset call
+     */
+    function getRemoveLiquiditySingleAssetCalldata(
+        address /* _setToken */,
+        address /*_pool*/,
+        address /*_component*/,
+        uint256 /*_minTokenOut*/,
+        uint256 /*_liquidity*/
+    )
+        external
+        pure
+        override
+        returns (address /*target*/, uint256 /*value*/, bytes memory /*data*/)
+    {
+        revert("Arrakis single asset removal is not supported");
+    }
+
+    /**
+     * Returns the address of the spender
+     */
+    function getSpenderAddress(address /*_pool*/)
+        external
+        view
+        override
+        returns (address spender)
+    {
+        spender = router;
+    }
+
+    /**
+     * Verifies that this is an Arrakis Vault pool holding valid UniswapV3 position
+     *
+     * @param  _pool          Address of liquidity token
+     * @param  _components    Address array of supplied/requested tokens
+     */
+    function isValidPool(address _pool, address[] memory _components)
+        external
+        view
+        override
+        returns (bool)
+    {
+        // Attempt to get the tokens of the provided pool
+        address token0;
+        address token1;
+        try IArrakisVaultV1(_pool).token0() returns (address _token0) {
+            token0 = _token0;
+        } catch {
+            return false;
+        }
+        try IArrakisVaultV1(_pool).token1() returns (address _token1) {
+            token1 = _token1;
+        } catch {
+            return false;
```

Comment:
I don't quite understand why we can't have this line covered, especially since line 234 seem to be covered.

Am I missing anything here ?

Reply:
Its not that we couldn't, but having a condition where the token0 external call succeeds and the token1 call fails would have to be done with a mock contract that exhibits this behavior.

Reply:
Ah I see. That explains why we can cover the first call easily but not the second one. 
I still think adding this little mock contract might be worth it if it helps us to just keep it at a nice 100% coverage.
Also might be worth looking into other mock options such as [smock](https://github.com/defi-wonderland/smock) ( or whatever the cool kids are using nowadays)

Reply:
I was scared of adding dependencies without upgrading existing dev deps, but I've heard of smock and would love to try it. If I run into an issue with configuring `smock`, I'll just do a small mock contract to get this to 100% which I agree is nice and why I added a constructor test :)